### PR TITLE
[sitecore-jss-cli] Add code extraction CLI command

### PR DESCRIPTION
**Pull request by art-alexeyenko**
- [x] This PR follows the [Contribution Guide](https://github.com/Sitecore/jss/blob/dev/CONTRIBUTING.md)
- [ ] Changelog updated
- [ ] Upgrade guide entry added

## Description / Motivation
Adds initial code extraction CLI command for future Design Library integration.
Notes:
- There will be a follow-up task to refine this, based on endpoints' implementation
- Pipeline fails at test step atm - will be fixed promptly.

## Testing Details
- [x] Unit Test Added
- [ ] Manual Test/Other (Please elaborate)

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: packages/sitecore-jss-cli/src/scripts/extract-components.ts [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    method: 'POST',
+    headers: {
+      Authorization: `Bearer ${token}`,
+      'Content-Type': 'application/json',
+    },
+    body: code,
+  });
+
+  if (!response.ok) {
+    console.error(chalk.red('Failed to send extracted code:', response.statusText));
```
Comment (yavorsk) [inline packages/sitecore-jss-cli/src/scripts/extract-components.ts]: It may be useful to add the componentPath to the success and error messages

File: packages/sitecore-jss-cli/src/scripts/extract-components.ts [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+  return yargs.command(
+    'extract-components',
+    'Reads component files and posts their code for AI learning',
+    args,
+    handler
+  );
+}
+
+/**
+ * Handler fo the extract-component API command
```
Comment (yavorsk) [inline packages/sitecore-jss-cli/src/scripts/extract-components.ts]:
```suggestion
 * Handler for the extract-component API command
```

File: packages/sitecore-jss-cli/src/scripts/extract-components.ts [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+  const tsHost = ts.createCompilerHost(tsOptions, true);
+  const sourceFile = tsHost.getSourceFile(componentBuilderPath, ts.ScriptTarget.Latest, (msg) => {
+    throw new Error(`Failed to parse ${componentBuilderPath}: ${msg}`);
+  });
+
+  const parseNodes = (node: ts.Node) => {
+    if (ts.isImportDeclaration(node)) {
+      const moduleName = node.moduleSpecifier.getText().replace(/['"]/g, '');
+      if (!moduleName.startsWith('node:') && moduleName.indexOf('/node_modules') === -1) {
+        const resolvedModule = ts.nodeModuleNameResolver(moduleName, __filename, tsOptions, tsHost);
```
Comment (yavorsk) [inline packages/sitecore-jss-cli/src/scripts/extract-components.ts]: I do not fully understand but Is it correct to use '__filename' here?

**Review by yavorsk — CHANGES_REQUESTED** [review]
looks good! see some comments

File: packages/sitecore-jss-cli/src/scripts/extract-components.ts [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    });
+    const jsonResponse = await authenticateResponse.json();
+    return jsonResponse.access_token;
+  } catch (error) {
+    console.error(chalk.red('Error authenticating with M2M token endpoint:', error));
+    return null;
+  }
+};
+
+export const resolveImportFiles = (appPath: string) => {
```
Comment (stasmaxymov) [inline packages/sitecore-jss-cli/src/scripts/extract-components.ts]: It is important for us not only extract component code, but associated name too. I think we can check a components  map for relevant names.

File: packages/sitecore-jss-cli/src/scripts/extract-components.ts [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+  return path.resolve(process.cwd(), appFolder);
+};
+
+export const fetchBearerToken = async () => {
+  try {
+    // TODO:adjust when M2M endpoint is live
+    const authenticateResponse = await fetch(M2M_ENDPOINT, {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/x-www-form-urlencoded',
```
Comment (stasmaxymov) [inline packages/sitecore-jss-cli/src/scripts/extract-components.ts]: content type is 'application/x-www-form-urlencoded' but body is JSON, does it work correctly?

File: packages/sitecore-jss-cli/src/scripts/extract-components.ts [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+}
+
+/**
+ * Handler fo the extract-component API command
+ * Reads imports from the componentBuilder.ts file and posts the code to the mesh endpoint
+ * @param {object} args - The arguments passed to the command, with optional appName string
+ * @returns {Promise<void>} void
+ */
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export async function handler(args: any) {
```
Comment (stasmaxymov) [inline packages/sitecore-jss-cli/src/scripts/extract-components.ts:46]: We will need a mechanism for verifying that run is happening on deploy, we don't want to over send components data each time. We will need to account for our XM deploy, Vercel, Netlify + way to configure it outside

File: packages/sitecore-jss-cli/src/scripts/extract-components.ts (hunk 3 above)
Comment (stasmaxymov) [inline packages/sitecore-jss-cli/src/scripts/extract-components.ts]: just to confirm, is this module resolver correctly resolves file locations even if import has different format, e.g. aliases like @/*

File: packages/sitecore-jss-cli/src/scripts/extract-components.ts [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+export const resolveImportFiles = (appPath: string) => {
+  const tsConfig = readConfigFile(path.resolve(appPath, 'tsconfig.json'), ts.sys.readFile);
+
+  if (tsConfig.error) {
+    throw new Error(`Error reading tsconfig.json from JSS app root: ${tsConfig.error.messageText}`);
+  }
+
+  const tsOptions = tsConfig.config.compilerOptions;
+
+  const componentBuilderPath = path.resolve(appPath, 'src', 'temp', 'componentBuilder.ts');
```
Comment (stasmaxymov) [inline packages/sitecore-jss-cli/src/scripts/extract-components.ts]: I think we should have a way to override a default location of component builder file

File: packages/sitecore-jss-cli/src/scripts/extract-components.ts [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+  const basePath = args.appFolder ? resolveAppPath(args.appFolder) : process.cwd();
+  try {
+    const componentPaths = resolveImportFiles(basePath);
+    const bearer = await fetchBearerToken();
+    if (!bearer) {
+      console.error(chalk.red('Failed to get bearer token, aborting code extraction'));
+      return;
+    }
+
+    for (const componentPath of componentPaths!) {
```
Comment (stasmaxymov) [inline packages/sitecore-jss-cli/src/scripts/extract-components.ts]: We probably should parallelize send requests, to speed up the process

**Review by stasmaxymov — CHANGES_REQUESTED** [review]

File: packages/sitecore-jss-cli/src/scripts/extract-components.ts (hunk 3 above)
Comment (art-alexeyenko) [reply packages/sitecore-jss-cli/src/scripts/extract-components.ts]: yes, this mechanism works with aliases, as long as `paths` are specified in compilerOptions retrieved from tsconfig

File: packages/sitecore-jss-cli/src/scripts/extract-components.ts (hunk 7 above)
Comment (art-alexeyenko) [reply packages/sitecore-jss-cli/src/scripts/extract-components.ts]: for v2 we would need that, true. For v1 it will always be in src/temp. With that in mind, do we still need it?

File: packages/sitecore-jss-cli/src/scripts/extract-components.ts (hunk 3 above)
Comment (art-alexeyenko) [reply packages/sitecore-jss-cli/src/scripts/extract-components.ts]: should be componentBuilderPath, you're right

**Review by illiakovalenko — COMMENTED** [review]
### File: `packages/sitecore-jss-cli/src/scripts/extract-components.ts`
- The script checks for `EXTRACT_CONSENT` before proceeding. This is a good safeguard, but it would be helpful to document this requirement in the CLI's documentation or usage instructions.
- The `M2M_ENDPOINT` and `meshEndpoint` URLs are hardcoded. While the `meshEndpoint` has a fallback to an environment variable, it would be better to make both endpoints configurable via environment variables.
- The `resolveImportFiles` function reads the `tsconfig.json` file to resolve imports. If the file is missing or invalid, it throws an error. The error message could suggest possible resolutions, such as checking the file path.
- The script includes TODO comments for future adjustments, such as refining the consent procedure and adjusting endpoints. These should be addressed before merging to ensure the feature is production-ready.

### File: `packages/sitecore-jss-cli/src/scripts/extract-components.test.ts`
- The test file provides good coverage for the `extract-components` script, including edge cases like missing environment variables and API failures.

### File: `package.json`
- The versions of the new dependencies are pinned. Consider using a caret (`^`) to allow for minor updates.

**Review by illiakovalenko — COMMENTED** [review]
### Suggested Code Changes to Resolve Comments

#### **File: `packages/sitecore-jss-cli/src/scripts/extract-components.ts`**

1. **Documenting `EXTRACT_CONSENT` Requirement**:
   Add a log message to inform users about the `EXTRACT_CONSENT` requirement.
   ```typescript
   // Add this log message at the start of the handler function
   console.log(chalk.yellow('Ensure that the EXTRACT_CONSENT environment variable is set to enable code extraction.'));
   ```

2. **Making Endpoints Configurable**:
   Replace the hardcoded `M2M_ENDPOINT` and `meshEndpoint` with environment variables.
   ```typescript
   const M2M_ENDPOINT = process.env.M2M_ENDPOINT || 'https://auth-staging-1.sitecore-staging.com/oauth/token';
   const meshEndpoint = `${process.env.SITECORE_EDGE_URL || constants.SITECORE_EDGE_URL_DEFAULT}/api/v1/mesh`;
   ```

3. **Improving Error Message for `resolveImportFiles`**:
   Update the error message to suggest possible resolutions.
   ```typescript
   if (tsConfig.error) {
     throw new Error(
       `Error reading tsconfig.json from JSS app root: ${tsConfig.error.messageText}. Please ensure the file exists and is valid.`
     );
   }
   ```

4. **Addressing TODO Comments**:
   Replace TODO comments with actionable items or remove them if no longer relevant.
   ```typescript
   // Example: Replace this TODO
   // TODO: adjust when mesh endpoint is live
   // With a more specific comment or remove it if already addressed
   ```

#### **File: `packages/sitecore-jss-cli/src/scripts/extract-components.test.ts`**

1. **Test Coverage**:
   No specific code changes are required here, but ensure that the test cases align with the updated functionality.

#### **File: `package.json`**

1. **Using Caret (`^`) for Dependency Versions**:
   Update the dependency versions to use caret (`^`) for minor updates.
   ```json
   "@types/proxyquire": "^1.3.31",
   "nock": "^14.0.0-beta.7",
   "proxyquire": "^2.1.3"
   ```

**Comment by illiakovalenko** [comment]
Hey @art-alexeyenko, quick general question: what do you think about moving all API requests, token resolution, and other non-CLI-related logic into sitecore-jss-dev-tools? Looks like now we have a lot within CLI
To follow the approach we are doing for the rest of commands (e.g. `deploy`)

**Comment by art-alexeyenko** [comment]
@illiakovalenko Yeah, that's not a bad idea. I left the functions within the cli command code cause they are only being used there. But in time we will reuse them.

File: packages/sitecore-jss-cli/src/scripts/extract-components.ts (hunk 4 above)
Comment (art-alexeyenko) [reply packages/sitecore-jss-cli/src/scripts/extract-components.ts]: We generally export the whole file content, but the component can also be imported as
`import { Component1, Component2 } from 'src/many-components';`
With this in mind, should we only use component file name when extracting?

File: packages/sitecore-jss-cli/src/scripts/extract-components.ts (hunk 7 above)
Comment (stasmaxymov) [reply packages/sitecore-jss-cli/src/scripts/extract-components.ts]: can users control the name of component factory by changing scripts? if not - we don't need this parameter

File: packages/sitecore-jss-cli/src/scripts/extract-components.ts (hunk 4 above)
Comment (stasmaxymov) [reply packages/sitecore-jss-cli/src/scripts/extract-components.ts]: Generally I suggest to parse `components.set('BYOCWrapper', BYOCWrapper);` expressions to find a proper name, as I understand we support scenario where components are declared statically in `scripts\generate-component-builder\plugins\components.ts` file and thus might have names that differ from file names.

As for named imports, don't we support it only for packages? I would say we need to account for named imports if we also support extracting components code from packages. E.g. if it is a monorepo setup. 
I would say that supporting extracting code from local packages, but kind of extra goal, and can be done as separate ticket

**Comment by illiakovalenko** [comment]
@art-alexeyenko Why not move it right now? We’re not reusing the 'deploy' - related methods - or any others, for that matter. So why do we need to keep it in "cli"?

File: packages/sitecore-jss-cli/src/scripts/extract-components.ts (hunk 4 above)
Comment (art-alexeyenko) [reply packages/sitecore-jss-cli/src/scripts/extract-components.ts]: Please check this new implementation

File: packages/sitecore-jss-cli/src/scripts/extract-components.ts (hunk 6 above)
Comment (art-alexeyenko) [reply packages/sitecore-jss-cli/src/scripts/extract-components.ts:46]: Done - I'll verify the XMC variables and adjust, but netlify and vercel logic is there

File: packages/sitecore-jss-dev-tools/src/codegen/utils.ts [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+            // get map.set assignments
+            mapAssignments.push(expressionNode);
+          }
+        }
+      });
+    } else if (ts.isExportAssignment(childNode)) {
+      // get component map export variable
+      mapExportName = childNode.expression.getText();
+    }
+    for (const mapAssignment of mapAssignments) {
```
Comment (stasmaxymov) [inline packages/sitecore-jss-dev-tools/src/codegen/utils.ts]: Shouldn't this `for` be outside of `ts.forEach`? You visit all nodes and populate complete map of imports and `.set` calls. Once you have complete info - you map component name to an import

Comment (art-alexeyenko) [reply packages/sitecore-jss-dev-tools/src/codegen/utils.ts]: indeed, you're right. fixed

**Review by stasmaxymov — APPROVED** [review]

**Comment by art-alexeyenko** [comment]
Closing this, recreating PR to merge into beta instead:
https://github.com/Sitecore/jss/pull/2075
